File: app/src/test/java/com/example/mike/mp3player/client/callbacks/SeekerBarController2Test.java
package com.example.mike.mp3player.client.callbacks;

import android.animation.ValueAnimator;
import android.content.Context;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import com.example.mike.mp3player.client.MediaControllerAdapter;
import com.example.mike.mp3player.client.views.SeekerBar;
import com.example.mike.mp3player.client.views.TimeCounter;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;
import org.robolectric.shadows.ShadowSeekBar;
import org.robolectric.shadows.ShadowValueAnimator;

import static android.support.v4.media.session.PlaybackStateCompat.STATE_PAUSED;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;

@RunWith(RobolectricTestRunner.class)
@Config(manifest=Config.NONE, sdk = 26, shadows = {ShadowValueAnimator.class, ShadowSeekBar.class})
public class SeekerBarController2Test {

    private Context m_context;
    @Mock
    private MediaControllerAdapter m_mediaControllerAdapter;

    private SeekerBarController2 m_seekerBarController2;
    private SeekerBar m_seekerBar;
    private final long DEFAULT_DURATION = 3000;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        m_context = RuntimeEnvironment.systemContext;
        m_seekerBar = new SeekerBar(m_context);
        m_seekerBarController2 = new SeekerBarController2(m_seekerBar, m_mediaControllerAdapter);
        // set default metadata
        m_seekerBarController2.onMetadataChanged(createMetaData(DEFAULT_DURATION));
    }

    @After
    public void tearDown() {
    }

    @Test
    public void testMetadataChanged() {
        ValueAnimator originalValueAnimator = m_seekerBarController2.getValueAnimator();
        final long DURATION = 1000;
        m_seekerBarController2.onMetadataChanged(createMetaData(DURATION));
        assertValueAnimatorReset(originalValueAnimator);
        ValueAnimator resultValueAnimator = m_seekerBarController2.getValueAnimator();
        assertEquals(DURATION, resultValueAnimator.getDuration());
    }

    @Test
    public void testPlaybackSpeedIncreaseStatePaused() {
        final float SPEED = 1.1f;
        testSpeedChange(SPEED);
    }

    @Test
    public void testStartTracking() {
        TimeCounter timeCounter = mock(TimeCounter.class);
        m_seekerBar.setTimeCounter(timeCounter);
        m_seekerBarController2.onStartTrackingTouch(m_seekerBar);
        assertTrue(m_seekerBar.isTracking());
    }

    @Test
    public void testStopTracking() {
//        TimeCounter timeCounter = mock(TimeCounter.class);
//        m_seekerBar.setTimeCounter(timeCounter);
        ValueAnimator valueAnimator = m_seekerBarController2.getValueAnimator();
        valueAnimator.start();
        m_seekerBarController2.onStopTrackingTouch(m_seekerBar);
        assertFalse(m_seekerBar.isTracking());
    }

    @Test
    public  void testPlaybackSpeedDecreasedSpeedPaused() {
        final float SPEED = 0.75f;
        testSpeedChange(SPEED);
    }

    private void testSpeedChange(float speed) {

        final long EXPECTED_DURATION = (long) (DEFAULT_DURATION / speed);
        final PlaybackStateCompat playbackState = createPlaybackState(STATE_PAUSED, 350, speed);
        final ValueAnimator originalValueAnimator = m_seekerBarController2.getValueAnimator();
        m_seekerBarController2.onPlaybackStateChanged(playbackState);
        assertValueAnimatorReset(originalValueAnimator);
        ValueAnimator valueAnimator = m_seekerBarController2.getValueAnimator();
        assertTrue(valueAnimator.isStarted());
        assertTrue(valueAnimator.isPaused());
        long resultDuration = valueAnimator.getDuration();
        String errorMessage = new StringBuilder().append("incorrect duration, expected: ")
                .append(EXPECTED_DURATION)
                .append(" but got ")
                .append(resultDuration)
                .toString();
        assertEquals(errorMessage, EXPECTED_DURATION, resultDuration);
    }

    private void assertValueAnimatorReset(ValueAnimator originalAnimator) {
        ValueAnimator resultValueAnimator = m_seekerBarController2.getValueAnimator();
        assertNotEquals("Value animator should have been recreated", resultValueAnimator, originalAnimator);

    }

    private MediaMetadataCompat createMetaData(long duration) {
        return new MediaMetadataCompat.Builder()
                .putLong(MediaMetadataCompat.METADATA_KEY_DURATION, duration)
                .build();
    }

    private PlaybackStateCompat createPlaybackState(int state, int position, float speed) {
        return new PlaybackStateCompat.Builder()
                .setState(state, position, speed)
                .build();
    }
}

File: client/src/test/java/com/github/goldy1992/mp3player/client/SeekerBarController2Test.kt
package com.github.goldy1992.mp3player.client

import android.animation.ValueAnimator
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.PlaybackStateCompat
import androidx.test.platform.app.InstrumentationRegistry
import com.github.goldy1992.mp3player.client.views.SeekerBar
import com.github.goldy1992.mp3player.client.views.TimeCounter
import com.nhaarman.mockitokotlin2.mock
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowSeekBar
import org.robolectric.shadows.ShadowValueAnimator

@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE, sdk = [26], shadows = [ShadowValueAnimator::class, ShadowSeekBar::class])
class SeekerBarController2Test {
    private var m_context: Context? = null

    private val m_mediaControllerAdapter: MediaControllerAdapter = mock<MediaControllerAdapter>()
    private lateinit var timeCounter: TimeCounter
    private lateinit var m_seekerBarController2: SeekerBarController2
    private lateinit var m_seekerBar: SeekerBar
    private val DEFAULT_DURATION: Long = 3000

    @Before
    fun setUp() {
        m_context = InstrumentationRegistry.getInstrumentation().context
        timeCounter = TimeCounter(Handler(Looper.getMainLooper()), m_mediaControllerAdapter)
        m_seekerBar = SeekerBar(m_context)
        m_seekerBarController2 = SeekerBarController2(m_mediaControllerAdapter, timeCounter)
        m_seekerBarController2.init(m_seekerBar)
        // set default metadata
        m_seekerBarController2.onMetadataChanged(createMetaData(DEFAULT_DURATION))
    }

    @After
    fun tearDown() {
    }

    @Test
    fun testMetadataChanged() {
        val originalValueAnimator = m_seekerBarController2!!.valueAnimator
        val DURATION: Long = 1000
        m_seekerBarController2!!.onMetadataChanged(createMetaData(DURATION))
        assertValueAnimatorReset(originalValueAnimator)
        val resultValueAnimator = m_seekerBarController2!!.valueAnimator
        Assert.assertEquals(DURATION, resultValueAnimator!!.duration)
    }

    @Test
    fun testPlaybackSpeedIncreaseStatePaused() {
        val SPEED = 1.1f
        testSpeedChange(SPEED)
    }

    @Test
    fun testStartTracking() {
        //    m_seekerBar.setTimeCounter(timeCounter);
        m_seekerBarController2.onStartTrackingTouch(m_seekerBar)
        Assert.assertTrue(m_seekerBar.isTracking)
    }

    @Test
    fun testStopTracking() { //        TimeCounter timeCounter = mock(TimeCounter.class);
//        m_seekerBar.setTimeCounter(timeCounter);
        val valueAnimator = m_seekerBarController2!!.valueAnimator
        valueAnimator!!.start()
        m_seekerBarController2!!.onStopTrackingTouch(m_seekerBar!!)
        Assert.assertFalse(m_seekerBar!!.isTracking)
    }

    @Test
    fun testPlaybackSpeedDecreasedSpeedPaused() {
        val SPEED = 0.75f
        testSpeedChange(SPEED)
    }

    private fun testSpeedChange(speed: Float) {
        val EXPECTED_DURATION = (DEFAULT_DURATION / speed).toLong()
        val playbackState = createPlaybackState(PlaybackStateCompat.STATE_PAUSED, 350, speed)
        val originalValueAnimator = m_seekerBarController2!!.valueAnimator
        m_seekerBarController2.onPlaybackStateChanged(playbackState)
        assertValueAnimatorReset(originalValueAnimator)
        val valueAnimator = m_seekerBarController2!!.valueAnimator
        Assert.assertTrue(valueAnimator!!.isStarted)
        Assert.assertTrue(valueAnimator.isPaused)
        val resultDuration = valueAnimator.duration
        val errorMessage = StringBuilder().append("incorrect duration, expected: ")
                .append(EXPECTED_DURATION)
                .append(" but got ")
                .append(resultDuration)
                .toString()
        Assert.assertEquals(errorMessage, EXPECTED_DURATION, resultDuration)
    }

    private fun assertValueAnimatorReset(originalAnimator: ValueAnimator?) {
        val resultValueAnimator = m_seekerBarController2!!.valueAnimator
        Assert.assertNotEquals("Value animator should have been recreated", resultValueAnimator, originalAnimator)
    }

    private fun createMetaData(duration: Long): MediaMetadataCompat {
        return MediaMetadataCompat.Builder()
                .putLong(MediaMetadataCompat.METADATA_KEY_DURATION, duration)
                .build()
    }

    private fun createPlaybackState(state: Int, position: Int, speed: Float): PlaybackStateCompat {
        return PlaybackStateCompat.Builder()
                .setState(state, position.toLong(), speed)
                .build()
    }
}